python/dgl/dataloading/fixed.py

```diff
@@ -0,0 +1,146 @@
+"""Fixed subgraph sampler."""
+from ..sampling.utils import EidExcluder
+from .base import set_node_lazy_features, set_edge_lazy_features, Sampler
+
+# import non-DGL libraries
+import numpy as np
+import torch
+from collections import defaultdict
+
+class FixedSampler(Sampler):
+    """Subgraph sampler that sets an upper bound on the number of nodes included in 
+    each layer of the sampled subgraph.
+    
+    At each layer, the frontier is randomly subsampled. Rare node types can also be 
+    upsampled by taking the scaled square root of the sampling probabilities.
+
+    It performs node-wise neighbor sampling and returns the subgraph induced by
+    all the sampled nodes.
+
+    Parameters
+    ----------
+    fanouts : list[int] or list[dict[etype, int]]
+        List of neighbors to sample per edge type for each GNN layer, with the i-th
+        element being the fanout for the i-th GNN layer.
+
+        If only a single integer is provided, DGL assumes that every edge type
+        will have the same fanout.
+
+        If -1 is provided for one edge type on one layer, then all inbound edges
+        of that edge type will be included.
+    fixed_k : int
+            The number of nodes to sample for each GNN layer.
+    upsample_rare_types : bool
+        Whether or not to upsample rare node types.
+    replace : bool, default True
+        Whether to sample with replacement
+    prob : str, optional
+        If given, the probability of each neighbor being sampled is proportional
+        to the edge feature value with the given name in ``g.edata``. The feature must be
+        a scalar on each edge.
+    """
+    def __init__(self, fanouts, fixed_k, upsample_rare_types, replace=False, prob=None, 
+                 prefetch_node_feats=None, prefetch_edge_feats=None, output_device=None):        
+        super().__init__()
+        self.fanouts = fanouts
+        self.replace = replace
+        self.fixed_k = fixed_k
+        self.upsample_rare_types = upsample_rare_types
+        self.prob = prob
+        self.prefetch_node_feats = prefetch_node_feats
+        self.prefetch_edge_feats = prefetch_edge_feats
+        self.output_device = output_device
+
+    def sample(self, g, seed_nodes, exclude_eids=None):
+        """Sampling function.
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The graph to sampler from.
+        seed_nodes : Tensor or dict[str, Tensor]
+            The nodes sampled in the current minibatch.
+        exclude_eids : Tensor or dict[etype, Tensor], optional
+            The edges to exclude from neighborhood expansion.
+
+        Returns
+        -------
+        input_nodes, output_nodes, subg
+            A triplet containing (1) the node IDs inducing the subgraph, (2) the node
+            IDs that are sampled in this minibatch, and (3) the subgraph itself.
+        """
+
+        # define empty dictionary to store reached nodes
```

Comment:
```suggestion
        # Define empty dictionary to store reached nodes.
```
nit: Same for other comments, treat it as an English sentence.

Reply:
Updated for all comments.